kyocera-ecosys-3500-4500: init at 9.0

me-and opened the pull request:
New package with drivers for various Kyocera printers.

Based heavily on the kyodialog package.

## Description of changes

New driver package for some Kyocera printers, based on the drivers from https://www.kyoceradocumentsolutions.com/download/model_en.html?r=119&s=25&m=391&p=83 and the kyodialog package at https://github.com/NixOS/nixpkgs/tree/master/pkgs/misc/cups/drivers/kyodialog

This is my first attempt at sharing a package; I have it building and working on my test NixOS VM, but I'm still getting the hang of the Nix language, and I'm not 100% certain of the changes I've made as part of rewriting this to work as a derivation rather than an import (or even if those are the right words for what I've done…)

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Pandapip1 reviewed (changes requested):
Please follow the instructions at https://github.com/NixOS/nixpkgs/actions/runs/9165436983/job/25198875210?pr=313244

Additionally, please add a commit that adds yourself as a maintainer.

Finally, printer drivers often have `cups` somewhere in their name. Can I suggest `cups-kyocera-ecosys-3500-4500` as the package name?

cole-h commented:
@ofborg eval

me-and committed:
maintainers: add me-and

me-and commented:
> Please follow the instructions at https://github.com/NixOS/nixpkgs/actions/runs/9165436983/job/25198875210?pr=313244

Done!

> Additionally, please add a commit that adds yourself as a maintainer.

Also done! I'd been wary of doing this as I didn't want to commit to significant ongoing work, but it looks like that's not likely in practice here.

> Finally, printer drivers often have `cups` somewhere in their name. Can I suggest `cups-kyocera-ecosys-3500-4500` as the package name?

I've gone for "cups-kyocera-3500-4500"; I'd initially started working on this to get drivers for my own Kyocera Ecosys printer, but the driver package includes drivers for several printers from different lines too. I have no idea what Kyocera's conventions are for naming their printers or packaging their drivers, but the common theme for this package seems to be them having 3500, 4000 or 4500 in the name.

me-and committed:
cups-kyocera-3500-4500: init at 9.0

New package with drivers for various Kyocera printers.

Based heavily on the kyodialog package.

Pandapip1 reviewed (approved):
Diff LGTM.

nicoonoclaste reviewed (changes requested):
LGTM aside from a minor refactor

me-and commented on pkgs/by-name/cu/cups-kyocera-3500-4500/package.nix:
```diff
@@ -0,0 +1,86 @@
+assert region == "Global" || region == "EU";
+
+let
+  kyodialog_version_short = "9";
```
Looking at the other Kyocera drivers, it seems the precedent is for them to use `9` here when the full version is `9.0`, and to use `9.1` when the full version is `9.1`. Given that, I'm inclined to leave this as is, and treat this as effectively an arbitrary string that'll need updating to a different arbitrary string for future releases, rather than implying it can be predictably computed.

That said, I don't have very strong feelings about this; I'm happy to make the change if that'd still be preferable.
